### ISSUE TITLE
Confused with the beginning of "stage 1"

### ISSUE THREAD

**Issue by pengpaiSH** [shuokay/resnet#3]
@shuokay Thank you for sharing your implementation of Residual Network in mxnet. When I read on your code, I have no idea why you treat the beginning of "stage 1" in separate codes.

```
    "stage 1"
    for i in xrange(3):
        if i == 0:
            _data = mx.symbol.Convolution(data=data,  num_filter=256,  kernel=(1,1), stride=(1,1), pad=(0,0))
            conv1 = ConvFactory(          data=data,  num_filter=64,   kernel=(1,1), stride=(1,1), pad=(0,0))
            conv2 = ConvFactory(          data=conv1, num_filter=64,   kernel=(3,3), stride=(1,1), pad=(1,1))
            conv3 = mx.symbol.Convolution(data=conv2, num_filter=256,  kernel=(1,1), stride=(1,1), pad=(0,0))
            data  = _data+conv3
            bn    = mx.symbol.BatchNorm(data=data)
            act   = mx.symbol.Activation(data=bn, act_type='relu')
            data  = act
```

Why not only `data =  ResidualFactory(data, (64, 64, 256), True)` just like stage 2-4?

Thank you for your time.

**Comment by shuokay** [comment]
Because there is no dimension change between the input data and "stage 1"

**Comment by pengpaiSH** [comment]
@shuokay Since there is no dimension change, why not directly use `data = ResidualFactory(data, (64, 64, 256), diff_dim=False)` ? It seems you would like to use a vanilla Convolutional Layer instead of `_data=data` ? Could you detail the reason, please ?

Again, thanks for your time!

**Comment by shuokay** [comment]
sorry, the the code is outdated, the example have moved to https://github.com/shuokay/mxnet/blob/master/example/image-classification/symbol_resnet-small.py

**Comment by pengpaiSH** [comment]
@shuokay Thank you for providing this updated "small" version. After reading your ResNet-Small code, I would like to ask:
(1) You hava a parameter `n` in your `ResidualSymbol()` function. Note that this residual network is totally different with your previously outdated version, which was a 152-layer as the original paper described.
(2) I guess you define such small residual network in order to tailor for the MNIST dataset. Because you define the class number is `10` and you follow the residual network by a Pooling layer with kernel (7,7). With dimension reduction (by 2) in "stage 2" and "stage 3", the output size is 7x7.
(3) With this specific small residual network, how much accuracy do you earn for the MNIST? And would you like to provide the learning rate, batch_size, etc in the `mx.model.FeedForward()`function, please?

**Comment by shuokay** [comment]
(1) ResNet-Small try to reproduce the result of cifar10 experiment in the resnet paper, `n` consist with that in resnet paper.
(2) I am using cifar10 dataset which has 10 classes. cifar10 is 32_32, mxnet's train_cifar10.py crop it to 28_28
(3) train accuracy about 91.75% and test accuracy about 84.30%. You can reproduce it running  
`python train_cifar10.py --network=resnet-small --lr=0.01 --lr-factor=0.1 --lr-factor-epoch=20 --num-epochs=30` 
in `mxnet/example/image-classification/`, you also can check the other hyper parameters in `train_model.py` and `train_cifar10.py`

**Comment by pengpaiSH** [comment]
@shuokay Thank you very much for you detailed answer, really appreciate that. I am now trying to reproduce the cifar-10 experiment with your ResNet code now. The paper claims that 110 layer ResNet achieves 6.63% error which seems promising. 

By the way, I am a green hand of mxnet (come from Keras community). It seems that the document is not that perfect. I would like to know how to automatically save the best model (the highest accuracy) during training process. Besides, is it possible to add a Early Stopping easily using callbacks?

Thank you!

**Comment by shuokay** [comment]
1. MxNet cant automatically save the best model currently.
2. MxNet didn't provide early stop, but I think you can add this feature by passing callback function to  `batch_end_callback` or `epoch_end_callback` of `mxnet.model.FeedForward`

**Comment by pengpaiSH** [comment]
@shuokay 
1.If MxNet cannot save the best model, how do you pick up the model after training precess finishes? 
2. Are you suggesting that I should write my own call_back function?

**Comment by pengpaiSH** [comment]
@shuokay 
Another question that I would like to ask is, in your resnet-small.py:

```
    data    = ConvFactory(data=mx.symbol.Variable(name='data'), num_filter=16, kernel=(3,3), stride=(1,1), pad=(1,1))
    res     = ResidualSymbol(data)
```

Before feeding in the ResNet, you are adding a ConvFactory which clearly is the one layer mentioned in the original paper "1+2n" layers. I don't not see any explanations for such extra single layer for? Would you please share your thoughts, please?

**Comment by shuokay** [comment]
The author have  written in the paper `the first layer is 3×3 convolutions`, I set `num_filter=16, kernel=(3,3), stride=1` to keep the dimension consistent.

**Comment by pengpaiSH** [comment]
@shuokay Thank you for your patience. One more practical question: since mxnet does not provide the functionality of automatically model saving. How do you pick up the model when the training process finishes?

**Comment by shuokay** [comment]
save the train log and analyse it.

**Comment by pengpaiSH** [comment]
@shuokay Do you mind sharing how to save the training log? Since the log shows on the screen automatically, it won't work if I type "python train.py >> log.txt".

**Comment by shuokay** [comment]
`python train.py &> log.txt` or pass a specific parameter to the `logging`(a little complicated compared to the former)

**Comment by pengpaiSH** [comment]
@shuokay Thanks for your tips~! I have one last question for your `resnet-small.py`. After the ResNet, you are ending with  a "global average pooling" layer with `kernel(7,7)` and a typical softmax.  Besides, I notice that you are using the same global average pooling layer for the out-dated resent.py. If you are trying to replace the typical fully-connected layer with the global average pooling as suggested in the paper "Network in Network"(I am just guessing). Would you please explain why you are using the kernel(7,7) ?

**Comment by shuokay** [comment]
Just ignore the old version resnet.py.
global average pooling means the the shape of pooling kernel is same to the entire feature map.
ps: there is an official resnet-28-small.py. https://github.com/dmlc/mxnet/pull/1230

**Comment by pengpaiSH** [comment]
@shuokay All right. Now I seem to understand. Say we have 128 feature maps and each size is `(7,7)`. Then,after conducting global average pooling you get a vector with size 128, right?

**Comment by shuokay** [comment]
absolutely right

**Comment by pengpaiSH** [comment]
@shuokay One more question. Say if I would like to draw a picture of training and validation accuracy over epochs using `matplotlib`, then `python train.py &> log.txt` is no necessary since I would like to save the accuracy into two lists `train_acc=[ ]` and `val_acc=[ ]`. Do you have any idea how to implement this by passing parameters into mxnet?

**Comment by shuokay** [comment]
There is a tool called parse_log.py in mxnet/tool/ but I am not sure is it what you need

**Comment by pengpaiSH** [comment]
@shuokay Thanks for telling me this parsing tool. Besides, I have found an example code of callbacks in the issues. 

```
def log_to_list(period, lst):
    def _callback(param):
        """The checkpoint function."""
        if param.nbatch % period == 0:
            name, value = param.eval_metric.get()
            lst.append(value)
    return _callback
```

I have tried it and it works for `batch_end_callbacks` instead of `epoch_end_callbacks`.
What if I would like to save the training accuracy and validation accuracy after each epoch?
![image](https://cloud.githubusercontent.com/assets/3724085/12316257/ab3961d0-babf-11e5-95d5-0494e1bb2093.png)
